rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,516 @@
+4. Compute the clone's checkpoint ID
+   1. If M is present, use M.`parent.parent_checkpoint` as the checkpoint ID
+   2. Otherwise, set checkpoint ID to a new v4 UUID
+6. If `parent_checkpoint` is not None, read it's manifest M_c. Otherwise let M_c be M_p
+7. Write a new manifest M'. If CAS fails, go to step 1. M' fields are set to:
+   - parent: set to point to checkpoint ID and initialized set to false
+   - writer_epoch: set to the epoch from M_c + 1.
+   - compactor_epoch: copied from M_c.
+   - wal_id_last_compacted: copied from M_c
+   - wal_id_last_seen: copied from M_c
+   - l0: copied from M_c
+   - l0_last_compacted: copied from M_c
+   - compacted: copied from M_c
+   - checkpoints: empty
+8. Create or update the checkpoint with checkpoint ID pointing to M_c in the parent DB. If CAS fails, go to step 1.
+9. Copy over any WAL SSTs between M_c.`last_compacted_wal_id` and M_c.`wal_id_last_seen`.
+10. Update M' with `initialized` set to true. If CAS fails, go to step 1.
+
+##### SST Path Resolution
+
+The DB may now need to look through multiple paths to find SST files, since it may need to open SST files written by a
+parent db. To support this, we’ll walk the manifest tree up to the root, and build a map from SST ID to Path at startup
+time, and maintain this in memory.
+
+##### Deleting a Database
+
+With the addition of clones, we need to be a bit more careful about how we delete a database. Previously we could
+simply remove it's objects from the Object Store and call it a day. With clones, it's possible that a SlateDB instance
+is holding a checkpoint that another database relies upon. So it's not safe to just delete that data. We'll instead
+require that users delete a database by calling the `Db::destroy` method. `Db::destroy` can specify that the delete is
+either a soft or hard delete by using the `soft` parameter. This way we can support both more conservative deletes that
+clean up data from the GC process after some grace period, and a simple path for hard-deleting a db.
+
+A soft delete will simply mark the manifest as deleted, and rely on a separate GC process to delete the db. It requires
+that the GC be running in some other long-lived process from the main writer. The GC process will wait for both some
+grace period, and for all checkpoints to either expire or be deleted. At that point it will delete the db's SSTs and
+exit. This is covered in more detail in the GC section. A soft delete will be processed by the writer as follows:
+1. Claim write-ownership of the DB by bumping the epoch and fencing the WAL. The intent here is to fence any active
+   writer. We assume that the writer will detect that it's fenced and exit by the time the GC goes to delete the data.
+2. Delete the writer's checkpoint, if any.
+3. Update the manifest by setting the `destroyed_at_s` field to the current time.
+
+A hard delete will proceed as follows. It is up to the user to ensure that there is no active writer at the time of
```

Comment:
shouldn't hard delete start with a soft delete first as well, since there can be race conditions?

Reply:
Yep. We should also block opening the db and creation of checkpoints on the presence of the the `destroyed_at_s` field. Otherwise we could fail mid-way through destroying the db and then permit a checkpoint to be created or the db to be opened. I'll update the RFC.